rust/aptos-indexer-transaction-stream/src/transaction_stream.rs

```diff
@@ -313,93 +309,150 @@ pub async fn create_fetcher_loop(
+        indexer_grpc_http2_ping_interval: Duration,
+        indexer_grpc_http2_ping_timeout: Duration,
+        indexer_grpc_reconnection_timeout_secs: Duration,
+        indexer_grpc_response_item_timeout_secs: Duration,
+        starting_version: u64,
+        request_ending_version: Option<u64>,
+        auth_token: String,
+        processor_name: String,
+        transaction_filter: TransactionFilter,
+        pb_channel_txn_chunk_size: usize,
+    ) -> Self {
+        Self {
+            indexer_grpc_data_service_address,
+            indexer_grpc_http2_ping_interval,
+            indexer_grpc_http2_ping_timeout,
+            indexer_grpc_reconnection_timeout_secs,
             indexer_grpc_response_item_timeout_secs,
-            resp_stream.next(),
+            starting_version,
+            request_ending_version,
+            auth_token,
+            processor_name,
+            transaction_filter,
+            pb_channel_txn_chunk_size,
+            resp_stream: None,
+            connection_id: None,
+            next_version_to_fetch: starting_version,
+            reconnection_retries: 0,
+            last_fetched_version: starting_version as i64 - 1,
+            fetch_ma: MovingAverage::new(3000),
+        }
+    }
+
+    pub async fn init_stream(&mut self) {
+        info!(
+            processor_name = self.processor_name,
+            service_type = PROCESSOR_SERVICE_TYPE,
+            stream_address = self.indexer_grpc_data_service_address.to_string(),
+            start_version = self.starting_version,
+            end_version = self.request_ending_version,
+            "[Parser] Connecting to GRPC stream",
+        );
+        let response = get_stream(
+            self.indexer_grpc_data_service_address.clone(),
+            self.indexer_grpc_http2_ping_interval,
+            self.indexer_grpc_http2_ping_timeout,
+            self.indexer_grpc_reconnection_timeout_secs,
+            self.starting_version,
+            self.request_ending_version,
+            self.auth_token.clone(),
+            self.processor_name.to_string(),
+        )
+        .await;
+        let connection_id = match response.metadata().get(GRPC_CONNECTION_ID) {
+            Some(connection_id) => connection_id.to_str().unwrap().to_string(),
+            None => "".to_string(),
+        };
+        self.connection_id = Some(connection_id);
+        self.resp_stream = Some(response.into_inner());
+        info!(
+            processor_name = self.processor_name,
+            service_type = PROCESSOR_SERVICE_TYPE,
+            stream_address = self.indexer_grpc_data_service_address.to_string(),
+            connection_id = self.connection_id,
+            start_version = self.starting_version,
+            end_version = self.request_ending_version,
+            "[Parser] Successfully connected to GRPC stream",
+        );
+    }
+
+    /// Gets a batch of transactions from the stream. Batch size is set in the grpc server.
```

Comment:
Again, maybe this specific PR isn't the right place for it, but we should really stop panicking everywhere. Let's make any function that could fail return a Result instead. This is the right time to make these changes, the SDK is our chance to clean up all our tech debt.

Reply:
To explain myself, panicking is vaguely acceptable when we're just doing it in our own service and we know it's the "top level" component of the system. The processor SDK could be used anywhere, it might just be one component of a larger system, we don't know how the user might use it. Our code shouldn't kill their entire process if it has an issue, we should let them handle it.